Implement LD approach for calculating weights

mumichae opened the pull request.

Moomboh committed:
add devcontainer configuration

Moomboh committed:
add R dependencies to devcontainer

Moomboh committed:
add failing testcase for 3 dataset multicca

Moomboh committed:
add pyomo dependencies to devcontainer

meritback committed:
untested model for lp approach

meritback committed:
datasets as tuples of tuples-> hashable

Commit:
added LinearApproach notebook

meritback committed:
linear approach objective function, init_w

meritback committed:
lasso and objective running, norm2 not working, not solving correctly

meritback committed:
nonlinear solver ipopt

Commit:
updated notebook

meritback committed:
solving now :)

meritback committed:
lp in sparsecca implemented

meritback committed:
multiple K

meritback committed:
Nerdbook with multiple k

meritback committed:
Nerdbook

Commit:
package structure changed so that tests uses the right function

Commit:
merge

Commit:
comment changes

Moomboh committed:
add ipopt to devcontainer

meritback committed:
implementation for K>1

yugeji commented on LinearApproach/LinearApproachMultiCCA.ipynb:
```diff
@@ -0,0 +1,567 @@
+{
```
<p>In the future, please import these functions from the existing package instead of copy pasting!!!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/theislab/sparsecca/pull/3/discussion/#comment-1111926239'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='7'/>

meritback replied:
<p>done:)</p>

yugeji commented on LinearApproach/LinearApproachMultiCCA.ipynb:
```diff
@@ -0,0 +1,567 @@
+{
```
<p>If you get the chance, please either change variable names to match Witten 2009 or specify your notation in a comment (which will be included in the docstring).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/theislab/sparsecca/pull/3/discussion/#comment-1111926251'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

meritback replied:
<p>changed PC to F, and samples to S, is that it?</p>

meritback replied:
<p>and model.Idx -&gt; model.N (N is number of matrices)</p>

yugeji commented on LinearApproach/LinearApproachMultiCCA.ipynb:
```diff
@@ -0,0 +1,567 @@
+{
```
<p>Add the mathematical equation for each constraint for readability.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/theislab/sparsecca/pull/3/discussion/#comment-1111926258'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='20'/>

rengesf replied:
<p>Done</p>

yugeji commented on LinearApproach/LinearApproachMultiCCA.ipynb:
```diff
@@ -0,0 +1,567 @@
+{
```
<p>I'm confused why this is necessary? Just a quick comment would be great</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/theislab/sparsecca/pull/3/discussion/#comment-1111926264'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

meritback replied:
<p>yes, this is creating the output from the model instance:</p><p>instance_non_linear.w_i_f.pprint() gives you:</p><pre class="ql-syntax" spellcheck="false">w_i_f : Size=10, Index=w_i_f_index
    Key    : Lower : Value                  : Upper : Fixed : Stale : Domain
    (0, 0) :     0 : -9.988716347647046e-09 :     1 : False :  True :  Reals
    (0, 1) :     0 :     0.5421267530639352 :     1 : False :  True :  Reals
    (0, 2) :     0 :     0.8402967294914276 :     1 : False :  True :  Reals
    (0, 3) :     0 : -9.979664917225373e-09 :     1 : False :  True :  Reals
    (0, 4) :     0 : -9.710092820626022e-09 :     1 : False :  True :  Reals
    (1, 0) :     0 :      0.290246540374087 :     1 : False :  True :  Reals
    (1, 1) :     0 :     0.9122425255472254 :     1 : False :  True :  Reals
    (1, 2) :     0 :     0.2890856799734783 :     1 : False :  True :  Reals
    (1, 3) :     0 : -9.969750471384377e-09 :     1 : False :  True :  Reals
    (1, 4) :     0 : -9.985112854439647e-09 :     1 : False :  True :  Reals
</pre><p>so we are basically just fetching the calculated values</p>

yugeji commented on LinearApproach/LinearApproachMultiCCA.ipynb:
```diff
@@ -0,0 +1,567 @@
+{
```
<p>Again, import instead please~</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/theislab/sparsecca/pull/3/discussion/#comment-1111926269'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='28'/>

meritback replied:
<p>so we just use the last line than? That is obviously always possible, ll the above can be done with <strong>iterative_process_K(datasets, K)</strong>, but I thought the idea of the notebook was to break things down? This is just not possible for the iterative process tho... should we delete K&gt;1 altogether from the notebook? Or just add<strong> iterative_process_K(datasets, K)</strong> as the first cell to tell users that this function does the below for K=1?</p>

meritback replied:
<p>updated in notebook</p><p> </p><p> </p>

yugeji commented on LinearApproach/LinearApproachMultiCCA.ipynb:
```diff
@@ -0,0 +1,567 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    def do_linear_approach(datasets, penalties):
</pre><p>Rename to something like <code>_update_w_lp</code> to distinguish it from <code>update_w</code>   and then make sure it has matching return types.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/theislab/sparsecca/pull/3/discussion/#comment-1111926278'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

meritback replied:
<p>They do not have the same output: </p><p>_update_w_lp: list, weights for all features for all matrices</p><p>update_w: w[f] float, weights for one feature and one matrix</p><p> </p>

yugeji commented on LinearApproach/LinearApproachMultiCCA.ipynb:
```diff
@@ -0,0 +1,567 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    def iterative_process_K(datasets:list, penalties:list, K:int):
</pre><p>Rename to something like <code>lp_pmd</code> o distinguish it from <code>multicca_pmd</code>  and then make sure it has matching return types. Also update the docstrings to match.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/theislab/sparsecca/pull/3/discussion/#comment-1111926286'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='33'/>

meritback replied:
<p>updated:) Hope it has the correct type?</p><p> </p><p> </p><p> </p>

yugeji commented on LinearApproach/LinearApproachMultiCCA.ipynb:
```diff
@@ -0,0 +1,567 @@
+{
```
<p>Since you've now added the code to the repo, please update this notebook so it no longer calls local functions, and instead demonstrates how to use <code>lp_pmd</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/theislab/sparsecca/pull/3/discussion/#comment-1111926291'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='35'/>

meritback replied:
<p>this will only be one call tho: w= lp_pmd(dataset, K) is this what you mean?</p>

meritback committed:
changed names of fuctions and output format of lp_pmd

Commit:
added mathematical equations and imported scale function

Commit:
updated test for la

meritback committed:
add multicca for testing

Commit:
__init__.py corrected

meritback committed:
notebook updated

meritback committed:
trying to finally push nerdbook

meritback committed:
lp_pmd called in nerdbook

Commit:
test works

Commit:
test works

meritback committed:
renamed variables as in Witten 2009

meritback committed:
:/ -> :(

meritback committed:
:)

meritback committed:
return also w_init

Commit:
updated tests/added correlation and fixed lp_pmd output

Commit:
uodated correlation computation

meritback committed:
new file copy of _multicca_pmd.py

meritback committed:
Merge branch 'lp_approach' of github.com:theislab-gobi-sc-dialogue/sparsecca into lp_approach

meritback committed:
copy of multicca_pmd file

Commit:
updated output

Commit:
import copy
import numpy as np
from ._multicca_pmd import multicca
from ._utils_pmd import scale

def multicca_permute(
        datasets: list,
        penalties: int = None,
        nperms: int = 10,
        niter: int = 3,
        standardize: bool = True,
):
    """
    Select tuning parameters for sparse multiple canonical correlation analysis
    using the penalized matrix decomposition.

     Params
    ------
    datasets: list of datasets
    niter : int (default: 3)
    standardize : bool (default: True)
        Whether to center and scale each dataset before computing sparse
        canonical variates.
    """
    datasets = copy.deepcopy(datasets)
    K = len(datasets)
    for k in range(K):
        if datasets[k].shape[0] < 2:
            raise ValueError('Need at least 2 features in each data set!')
        if standardize:
            datasets[k] = scale(datasets[k], center=True, scale=True)

    if penalties is None:
        # only "standard" mode
        penalties = np.zeros(shape=(K, 10))
        for k in range(K):
            n_cols = datasets[k].shape[1]
            vals = np.linspace(start=.1, stop=.8, num=10) * np.sqrt(n_cols)
            penalties[k, :] = [np.max([v, 1.1]) for v in vals]
    else:
        penalties = np.full(shape=(K, 1), fill_value=penalties)

    penalty_columns = penalties.shape[1]
    cors = np.zeros(shape=penalty_columns)
    for i in range(penalty_columns):
        ws, ws_init = multicca(datasets, penalties=penalties[:, i], niter=niter)
        cors[i] = get_cors(datasets, ws)

    # permute
    perm_cors = np.zeros(shape=(nperms, penalty_columns))
    for j in range(nperms):
        datasets_perm = datasets

        # permute the data
        for k in range(K):
            datasets_perm[k] = np.random.permutation(datasets_perm[k])

        # run MultiCCA
        for i in range(penalty_columns):
            ws, _ = multicca(datasets, penalties=penalties[:, i], niter=niter)
            perm_cors[j, i] = get_cors(datasets, ws)

    # get summary statistics
    p_vals = []
    z_stats = []
    for i in range(penalty_columns):
        cor = cors[i]
        perm_cor = perm_cors[:, i]

        p_value = np.mean(perm_cor >= cor)
        p_vals.append(p_value)

        z_stat = (cor - np.mean(perm_cor)) / (np.std(perm_cor) + .05)
        z_stats.append(z_stat)

    return dict(
        pvals=np.array(p_vals),
        zstat=np.array(z_stats),
        bestpenalties=penalties[:, np.argmax(z_stats)],
        cors=np.array(cors),
        penalties=penalties,
        nperms=nperms
    )

def get_cors(datasets, ws):
    """
    Get sum of pairwise correlations
    """
    K = len(datasets)
    cors = []
    for i in range(1, K):
        for j in range(i):
            x = np.ravel(datasets[i] @ ws[i])
            y = np.ravel(datasets[j] @ ws[j])
            cor = np.corrcoef(x, y)
            cors.append(cor[0, 1])
    return sum(cors)

meritback committed:
initialized weights with svd

Commit:
Comparing Permuation of LA to R function

meritback committed:
Xn order has impact on weights in pmd approach, but not in lp approach

meritback committed:
lp>>pmd

Commit:
comparison la and pmd now also possible for k>1

yugeji commented on tests/sparsecca/sparsecca/_multicca_lp.py:
```diff
@@ -0,0 +1,172 @@
+# Linear Programming 
+
+
+def preprocess_datasets(datasets:list, standardize=True, mimic_R=True):
```
I mentioned this already, but this is copy and pasted :.( if you find yourself using something from `_multicca_pmd`, that just means that function should be a utility function and it should be moved accordingly.

yugeji commented on tests/sparsecca/sparsecca/_multicca_lp.py:
```diff
@@ -0,0 +1,172 @@
+    w_init = [np.full((len(datasets[0][0]),K), 0.5)]*len(datasets)
+    return w_final, w_init
+
+def multicca(datasets:list, penalties:list,  K:int, standardize, mimic_R):
```
No need for this!!

yugeji reviewed (commented):
For the most part, everything looks great! Just one major comment - in the future, if you're moving a git tracked file around, use `git mv` instead of `mv`. This allows git to track the file - right now, git thinks you deleted some files and created some new ones. Also, it's good practice to completely avoid duplication, either in files, functions, or whatever.
